UM/Resources.py

```diff
@@ -264,55 +268,188 @@ def __find(cls, resource_type: int, *args) -> List[str]:
+        # use nested structure: <app-name>/<version>/...
+        if cls.ApplicationVersion == "master" or cls.ApplicationVersion == "unknown":
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, cls.ApplicationVersion)
+        else:
+            from UM.Version import Version
+            version = Version(cls.ApplicationVersion)
+            storage_dir_name = os.path.join(cls.ApplicationIdentifier, "%s.%s" % (version.getMajor(), version.getMinor()))
+
+        # config is saved in "<CONFIG_ROOT>/<storage_dir_name>"
+        cls.__config_storage_path = os.path.join(Resources.getConfigStorageRootPath(), storage_dir_name)
+
+        # data is saved in
+        #  - on Linux: "<DATA_ROOT>/<storage_dir_name>"
+        #  - on other: "<CONFIG_DIR>" (in the config directory)
+        data_root_path = Resources.getDataStorageRootPath()
+        cls.__data_storage_path = cls.__config_storage_path if data_root_path is None else \
+            os.path.join(data_root_path, storage_dir_name)
+
+        # cache is saved in
+        #  - on Linux:   "<CACHE_DIR>/<storage_dir_name>"
+        #  - on Windows: "<CACHE_DIR>/<storage_dir_name>/cache"
+        #  - on Mac:     "<CONFIG_DIR>/cache" (in the config directory)
+        cache_root_path = Resources.getCacheStorageRootPath()
+        if cache_root_path is None:
             cls.__cache_storage_path = os.path.join(cls.__config_storage_path, "cache")
+        else:
+            cls.__cache_storage_path = os.path.join(cache_root_path, storage_dir_name)
+            if platform.system() == "Windows":
+                cls.__cache_storage_path = os.path.join(cls.__cache_storage_path, "cache")
+
+        if not os.path.exists(cls.__config_storage_path):
+            cls._copyLatestDirsIfPresent()
 
         cls.__paths.insert(0, cls.__data_storage_path)
 
+    ##  Copies the directories of the latest version on this machine if present, so the upgrade will use the copies
+    #   as the base for upgrade
+    @classmethod
+    def _copyLatestDirsIfPresent(cls):
+        # paths for the version we are running right now
+        this_version_config_path = Resources.getConfigStoragePath()
+        this_version_data_path = Resources.getDataStoragePath()
+
+        # find the latest existing directories on this machine
+        config_root_path_list = Resources.getPossibleConfigStorageRootPathList()
+        data_root_path = Resources.getDataStorageRootPath()
+
+        latest_config_path = Resources._findLatestStorageDirInPaths(config_root_path_list)
+        if not latest_config_path:
+            # no earlier storage dirs found, do nothing
+            return
+
+        if latest_config_path == this_version_config_path:
+            # if the directory found matches the current version, do nothing
+            return
+
+        # we first copy everything to a temporary folder, and then move it to the new folder
+        def _copyDirToDirViaTempDir(src_path, dest_path):
```

Comment:
The closure can go. It just means that this function is not meant for anything out of this scope.